File: UpdateBoard.java
// Time Complexity : The time complexity is O(m*n) where m is the number of rows and n is the number of column
// Space Complexity : The space complexity is O(m*n) where m is the number of rows and n is the number of column
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {

    int[][] dirs = {{1,0},{0,1},{-1,0},{0,-1},{-1,-1},{1,1},{1,-1},{-1,1}};

    public char[][] updateBoard(char[][] board, int[] click) {

        int rows = board.length;
        int columns = board[0].length;

        Queue<Integer> q = new LinkedList<>();
        q.offer(click[0]);
        q.offer(click[1]);

        if(board[click[0]][click[1]] == 'E'){
            board[click[0]][click[1]] = 'B';
        }

        if(board[click[0]][click[1]] == 'M') {
            board[click[0]][click[1]] = 'X';
            return board;
        }

        while(!q.isEmpty()){

            int row = q.poll();
            int column = q.poll();

            int count = mines(board,row,column);

            //if the cell has adjacent mines
            if(count != 0){
                board[row][column] = (char)(count+'0');
                continue;
            }

            //if the cell doesnt have adjacent mines
            for(int[] dir:dirs){

                int nextRow = row + dir[0];
                int nextColumn = column + dir[1];

                if(nextRow>=0 && nextRow<rows && nextColumn>=0 && nextColumn<columns){
                    if(board[nextRow][nextColumn] == 'E'){
                        q.offer(nextRow);
                        q.offer(nextColumn);
                        board[nextRow][nextColumn] = 'B';
                    }
                }
            }
        }
        return board;
    }

    public int mines(char[][] board,int row,int column){

        int count = 0;
        int rows = board.length;
        int columns = board[0].length;

        for(int[] dir:dirs){

            int nextRow = row + dir[0];
            int nextColumn = column + dir[1];

            if(nextRow>=0 && nextRow<rows && nextColumn>=0 && nextColumn<columns){
                if(board[nextRow][nextColumn] == 'M'){
                    count++;
                }
            }
        }

        return count;
    }
}

File: updateBoard.py
# Time Complexity : O(v+e) where v is the vertices of the graph and e is the edges
# Space Complexity : O(v+e)
# Did this code successfully run on Leetcode : Yes
class Solution:
    def updateBoard(self, board: List[List[str]], click: List[int]) -> List[List[str]]:

        if not board or not board[0]:
            return None

        x = click[0]
        y = click[1]

        if board[x][y] == 'M':
            board[x][y] = 'X'
            return board

        q = collections.deque([(x, y)])
        OFFSETS = [(-1, -1), (-1, 0), (-1, 1), (0,1), (1,1), (1,0), (1, -1), (0, -1)]

        while q:
            r, c = q.popleft()

            if board[r][c] == 'E':
                count = 0
                for delta_r, delta_c in OFFSETS:
                    next_r, next_c = r + delta_r, c + delta_c
                    if 0 <= next_r < len(board) and 0 <= next_c < len(board[0]):
                        if board[next_r][next_c] == 'M':
                            count += 1
                if count != 0:
                    board[r][c] = str(count)
                else:
                    board[r][c] = 'B'
                    for delta_r, delta_c in OFFSETS:
                        next_r, next_c = r + delta_r, c + delta_c
                        if 0 <= next_r < len(board) and 0 <= next_c < len(board[0]) and board[next_r][next_c] == 'E':
                            q.append((next_r, next_c))

        return board
